core/git_command.py

```diff
@@ -150,19 +156,66 @@ def raise_error(msg):
                     stderr.decode()
                 )
 
-        if not p.returncode == 0 and throw_on_stderr:
-            raise_error("`{}` failed with following output:\n{}\n{}".format(
-                command_str, stdout, stderr
-            ))
+        def on_result(return_code, stdout, stderr):
+            if decode and not isinstance(stdout, str):
+                stdout, stderr = self.decode_stdout(stdout, savvy_settings), stderr.decode()
 
-        if show_panel:
-            if savvy_settings.get("show_input_in_output"):
-                util.log.panel("> {}\n{}\n{}".format(command_str, stdout, stderr))
-            else:
-                util.log.panel("{}\n{}".format(stdout, stderr))
+            if return_code != 0 and throw_on_stderr:
+                raise_error("`{}` failed  ({}) with following output:\n{}\n{}".format(
+                    command_str, return_code, stdout, stderr
+                ))
+
+            if show_panel:
+                if savvy_settings.get("show_input_in_output"):
+                    util.log.panel("> {}\n{}\n{}".format(command_str, stdout, stderr))
+                else:
+                    util.log.panel("{}\n{}".format(stdout, stderr))
 
+        if not p.returncode == 0 and throw_on_stderr:
+            # this piece of code detect git error meesages related to authentification.
+            # stdin must be None, because we will use stdin to pass username and password from Sublime to Git.
+            if catch_credential_errors and stdin is None and 'fatal: could not read Username for' in stderr:
+                command = [self.git_binary_path, '-C', working_dir or self.repo_path] + list(arg for arg in args if arg)
+                command_str = " ".join(command)
+
+                self._stdout = None
+                is_window_command = hasattr(self, "window")
+                window = self.window if is_window_command else self.view.window()
+                window.show_input_panel('username:', '', lambda username: self.on_username(window, command_str, username, on_result), None, self.on_cancel)
+
+                # wait that the Git command has been executed in the show_input_panel thread.
+                if is_window_command:
+                    while self._stdout is None:
+                        time.sleep(1)
```

Comment:
Here I'm waiting the user to input his login and password. This is only possible with a `WindowCommand`. In case of a `TextCommand` I return immediately, meaning that the caller won't have access to the stdout.
This is generally not a big deal since nobody uses the output of "push", "pull" or "fetch".